pysdsl/types/encodedvector.hpp

```diff
@@ -148,33 +148,85 @@ template <class KEY_T>
 inline std::string key_to_string(KEY_T key) { return std::to_string(key); }
 
 
-template <class Sequence, typename KEY_T>
-inline
-auto add_dac_vector(py::module& m, KEY_T key,
-                    const char* doc = nullptr)
+namespace {
+
+const char dprrr[] = "DPRRR";
+const char dp[] = "DP";
+
+template <typename T, T t, bool = std::is_integral<T>::value>
+struct get_vector_type {};
+
+template <typename T, T t>
+struct get_vector_type<T, t, true> : sdsl::dac_vector<t> {};
+
+template <>
+struct get_vector_type<const char*, dp, false> : sdsl::dac_vector_dp<> {};
+
+template <>
+struct get_vector_type<const char*, dprrr, false> : sdsl::dac_vector_dp<sdsl::rrr_vector<>> {};
+
+} // namespace
+
+
+class add_dac_vector_functor
 {
-    auto name = "DirectAccessibleCodesVector" + key_to_string(key);
+public:
+    constexpr add_dac_vector_functor(py::module& m, const char* doc = nullptr,
+                                                    const char* doc_dp = nullptr): 
+        m(m), doc(doc), doc_dp(doc_dp) {}
 
-    auto cls = py::class_<Sequence>(m, name.c_str()).def(py::init());
 
-    add_sizes(cls);
-    add_description(cls);
-    add_serialization(cls);
-    add_to_string(cls);
+    template <typename KEY_T, KEY_T key>
+    inline
+    decltype(auto) get_vector(std::integral_constant<KEY_T, key>) {
+        using type = get_vector_type<KEY_T, key>;
+        auto name = "DirectAccessibleCodesVector" + key_to_string(key);
 
-    add_read_access<Sequence>(cls);
-    add_std_algo<Sequence>(cls);
+        auto cls = py::class_<type>(m, name.c_str()).def(py::init());
 
-    if (doc) {
-        cls.doc() = doc; }
+        add_sizes(cls);
+        add_description(cls);
+        add_serialization(cls);
+        add_to_string(cls);
```

Comment:
... like those

Reply:
But when later we auto-define constructors for them it doesn't work, because constructors are not inherited.
As a result `enable_if... is_constructible` in creation.hpp don't accept you classes forcing slow construction with intermediate classes (which is not implemented)